File: evals/behavioral/tasks/rust/fixtures/conn-protocol.bad.rs
// Runtime state: one struct holds a state enum and every operation is always
// callable, deferring "wrong state" to a runtime check (or nothing). So
// `Conn::closed().send(..)` type-checks — the protocol is not enforced by types.
pub enum State {
    Closed,
    Listening,
    Established,
}

pub struct Conn {
    state: State,
}

impl Conn {
    pub fn closed() -> Conn {
        Conn { state: State::Closed }
    }
    pub fn listen(mut self) -> Conn {
        self.state = State::Listening;
        self
    }
    pub fn accept(mut self) -> Conn {
        self.state = State::Established;
        self
    }
    pub fn send(self, _data: &str) -> Conn {
        self
    }
}

File: evals/behavioral/tasks/typescript/fixtures/conn-protocol.bad.ts
// Runtime state: one class holds a state field and every operation is always
// callable, deferring "wrong state" to a runtime throw (or nothing). So
// `closedConn().send(..)` type-checks — the protocol is not enforced by types.
type State = "closed" | "listening" | "established";

export class Conn {
  private state: State = "closed";

  listen(): Conn {
    if (this.state !== "closed") throw new Error("not closed");
    this.state = "listening";
    return this;
  }
  accept(): Conn {
    if (this.state !== "listening") throw new Error("not listening");
    this.state = "established";
    return this;
  }
  send(_data: string): Conn {
    if (this.state !== "established") throw new Error("not established");
    return this;
  }
}

export function closedConn(): Conn {
  return new Conn();
}
